### PR TITLE
stage5: changes according to the stage5

### PR DESCRIPTION
**Pull request by TatianaKoma**

### CODE REVIEW

File: src/ua/griddynamics/tictactoe/Game.java [hunk 1]
```diff
@@ -10,6 +10,16 @@ public Game(String input) {
         this.input = input;
     }
 
+    public static boolean isGridValid(String grid) {
```
Comment (isubo) [inline src/ua/griddynamics/tictactoe/Game.java:13]: Is it used?

File: src/ua/griddynamics/tictactoe/Game.java [hunk 2]
```diff
@@ -10,6 +10,16 @@ public Game(String input) {
...
+        int count = 0;
+        for (int i = 0; i < grid.length(); i++) {
+            if (grid.charAt(i) == '_') {
+                count++;
+            }
+        }
+        return count == 9;
+    }
+
     public void showGrid() {
```
Comment (isubo) [inline src/ua/griddynamics/tictactoe/Game.java]: Let's implement a function that returns string

File: src/ua/griddynamics/tictactoe/Main.java [hunk 3]
```diff
@@ -4,14 +4,31 @@
...
-
         Scanner scanner = new Scanner(System.in);
-        System.out.println("Enter cells: ");
-        String input = scanner.nextLine();
-
-        Game game = new Game(input);
+        Game game = new Game("_________");
         game.showGrid();
 
+        while (!game.isWinner('X') && !game.isWinner('O') && game.isNotFinished()) {
```
Comment (isubo) [inline src/ua/griddynamics/tictactoe/Main.java]: instead of calling separate methods isNotFinished can check if somebody wined or there is a draw

File: src/ua/griddynamics/tictactoe/Game.java (hunk 1 above)
Comment (isubo) [inline src/ua/griddynamics/tictactoe/Game.java]: Grid dimension should be a constant.

File: src/ua/griddynamics/tictactoe/Game.java [hunk 4]
```diff
@@ -26,34 +36,16 @@ public void showGrid() {
...
-            if (input.charAt(i) == O) {
-                countO++;
-            }
-        }
-        return Math.abs(countX - countO) >= IMPOSSIBLE_DIFFERENCE;
-    }
-
-    private boolean isWinner(char symbol) {
+    public boolean isWinner(char symbol) {
         int[][] winIndexes = {{0, 1, 2}, {3, 4, 5},
```
Comment (isubo) [inline src/ua/griddynamics/tictactoe/Game.java]: Looks like winIndexes  should be a constant

File: .gitignore [hunk 5]
```diff
@@ -21,3 +21,5 @@
 
 # virtual machine crash logs, see http://www.java.com/en/download/help/error_hotspot.xml
 hs_err_pid*
+.idea/
+out/
```
Comment (isubo) [inline .gitignore]: Add empty line at the end of the file

File: src/ua/griddynamics/tictactoe/Coordinate.java [hunk 6]
```diff
@@ -1,35 +1,49 @@
 package ua.griddynamics.tictactoe;
 
 public class Coordinate {
-    private int x;
-    private int y;
+    private final int x;
+    private final int y;
+    static int MIN_RANGE = 1;
```
Comment (isubo) [inline src/ua/griddynamics/tictactoe/Coordinate.java]: Coordinate should not know anything about min and max range.

File: src/ua/griddynamics/tictactoe/Game.java [hunk 7]
```diff
@@ -93,23 +63,26 @@ public void putAnswer(char symbol, Coordinate coordinate) {
...
-            case 2:
-                position = 2;
-                break;
-            case 3:
-                position = 5;
-                break;
-        }
-        return position + coordinate.getY();
+        return switch (coordinate.getX()) {
+            case 1 -> coordinate.getY() - coordinate.getX();
```
Comment (isubo) [inline src/ua/griddynamics/tictactoe/Game.java]: 1,2,3 -  it looks like some magic numbers
Add a new property `size` to Game. Use for loop instead of calling each X-value

File: src/ua/griddynamics/tictactoe/Game.java [hunk 8]
```diff
@@ -93,23 +63,26 @@ public void putAnswer(char symbol, Coordinate coordinate) {
...
+        };
     }
 
     public boolean isEmpty(Coordinate coordinate) {
         int position = getPosition(coordinate);
-        return input.charAt(position) == '_';
+        return input.charAt(position) == EMPTY;
+    }
+
+    public String showResult() {
```
Comment (isubo) [inline src/ua/griddynamics/tictactoe/Game.java]: showResult -> getResult

File: src/ua/griddynamics/tictactoe/Game.java [hunk 9]
```diff
@@ -1,68 +1,42 @@
...
+            {1, 4, 7}, {2, 5, 8},
+            {0, 4, 8}, {2, 4, 6}};
 
     public Game(String input) {
         this.input = input;
     }
 
-    public void showGrid() {
-        System.out.println("---------");
+    public String showGrid() {
```
Comment (isubo) [inline src/ua/griddynamics/tictactoe/Game.java]: showGrid -> getGrid

File: src/ua/griddynamics/tictactoe/Game.java [hunk 10]
```diff
@@ -1,68 +1,42 @@
...
+        StringBuilder sb = new StringBuilder();
+        sb.append("---------");
+        sb.append(System.getProperty("line.separator"));
```
Comment (isubo) [inline src/ua/griddynamics/tictactoe/Game.java]: Use `System.lineSeparator()` instead.
It is better to avoid some strings if there is already constant exists

File: src/ua/griddynamics/tictactoe/Game.java [hunk 11]
```diff
@@ -1,68 +1,42 @@
...
-                {1, 4, 7}, {2, 5, 8},
-                {0, 4, 8}, {2, 4, 6}};
-
-        for (int i = 0; i < winIndexes.length; i++) {
-            boolean firstWinIndex = input.charAt(winIndexes[i][0]) == symbol;
-            boolean secondWinIndex = input.charAt(winIndexes[i][1]) == symbol;
-            boolean thirdWinIndex = input.charAt(winIndexes[i][2]) == symbol;
+    public boolean isWinner(char symbol) {
+        for (int[] win_index : WIN_INDEXES) {
+            boolean firstWinIndex = input.charAt(win_index[0]) == symbol;
```
Comment (isubo) [inline src/ua/griddynamics/tictactoe/Game.java]: use for loop instead

File: src/ua/griddynamics/tictactoe/Main.java [hunk 12]
```diff
@@ -4,48 +4,39 @@
...
             }
-            coordinate = Coordinate.parse(strAnswer);
-
-            if (!coordinate.isInRange()) {
-                System.out.println("Coordinates should be from 1 to 3");
-                isNotInRange = true;
-                continue;
-            } else {
-                isNotInRange = false;
+            if (game.isFinished()) {
```
Comment (isubo) [inline src/ua/griddynamics/tictactoe/Main.java:16]: Game can not be finished on this step. That is why this check is not needed

File: src/ua/griddynamics/tictactoe/Main.java [hunk 13]
```diff
@@ -4,48 +4,39 @@
...
             }
         }
+        System.out.println(game.showResult());
+    }
 
-        game.putAnswer('X', coordinate);
-        game.showGrid();
+    private static void playStep(Scanner scanner, Game game, char symbol) {
+        Coordinate coordinate = null;
+        String strAnswer;
```
Comment (isubo) [inline src/ua/griddynamics/tictactoe/Main.java]: Why strAnswer is not inside do while block?

File: src/ua/griddynamics/tictactoe/readme.md [hunk 14]
```diff
@@ -0,0 +1,125 @@
+<img src="icon.png" align="right" />
```
Comment (isubo) [inline src/ua/griddynamics/tictactoe/readme.md]: Why was this file added?

File: src/ua/griddynamics/tictactoe/Game.java [hunk 15]
```diff
@@ -1,68 +1,42 @@
 package ua.griddynamics.tictactoe;
 
 public class Game {
-    private static final int IMPOSSIBLE_DIFFERENCE = 2;
     private final char X = 'X';
     private final char O = 'O';
+    private final char EMPTY = '_';
     private String input;
+    int[][] WIN_INDEXES = {{0, 1, 2}, {3, 4, 5},
```
Comment (isubo) [inline src/ua/griddynamics/tictactoe/Game.java]: should it be a `private static final`  variable?

File: .gitignore (hunk 5 above)
Comment (TatianaKoma) [reply .gitignore]: done

File: src/ua/griddynamics/tictactoe/Game.java (hunk 15 above)
Comment (TatianaKoma) [reply src/ua/griddynamics/tictactoe/Game.java]: Done

File: src/ua/griddynamics/tictactoe/readme.md (hunk 14 above)
Comment (TatianaKoma) [reply src/ua/griddynamics/tictactoe/readme.md]: Done

File: src/ua/griddynamics/tictactoe/Game.java (hunk 10 above)
Comment (TatianaKoma) [reply src/ua/griddynamics/tictactoe/Game.java]: Done

File: src/ua/griddynamics/tictactoe/Game.java (hunk 9 above)
Comment (TatianaKoma) [reply src/ua/griddynamics/tictactoe/Game.java]: Done

File: src/ua/griddynamics/tictactoe/Game.java (hunk 8 above)
Comment (TatianaKoma) [reply src/ua/griddynamics/tictactoe/Game.java]: Done

File: src/ua/griddynamics/tictactoe/Coordinate.java (hunk 6 above)
Comment (TatianaKoma) [reply src/ua/griddynamics/tictactoe/Coordinate.java]: Done

File: src/ua/griddynamics/tictactoe/Game.java (hunk 7 above)
Comment (TatianaKoma) [reply src/ua/griddynamics/tictactoe/Game.java]: Done

File: src/ua/griddynamics/tictactoe/Game.java (hunk 11 above)
Comment (TatianaKoma) [reply src/ua/griddynamics/tictactoe/Game.java]: Done

File: src/ua/griddynamics/tictactoe/Main.java (hunk 13 above)
Comment (TatianaKoma) [reply src/ua/griddynamics/tictactoe/Main.java]: Done

File: src/ua/griddynamics/tictactoe/Game.java [hunk 16]
```diff
@@ -1,88 +1,69 @@
...
-    private final char O = 'O';
+    public static final char X = 'X';
+    public static final char O = 'O';
+    public static final char EMPTY = '_';
+    private static final int SIZE_OF_GRID = 3;
+    private static final int[][] WIN_INDEXES = {{0, 1, 2}, {3, 4, 5},
+            {6, 7, 8}, {0, 3, 6},
+            {1, 4, 7}, {2, 5, 8},
+            {0, 4, 8}, {2, 4, 6}};
     private String input;
```
Comment (isubo) [inline src/ua/griddynamics/tictactoe/Game.java]: It is bette to use array instead

File: src/ua/griddynamics/tictactoe/Game.java [hunk 17]
```diff
@@ -1,88 +1,69 @@
...
-            }
-        }
-        return false;
+    public static boolean isDigit(String[] coordinates) {
+        return coordinates.length == 2 && coordinates[0].matches("\\d+") && coordinates[1].matches("\\d+");
     }
 
-    private boolean isTwoWinners() {
-        return isWinner(X) && isWinner(O);
+    public static boolean isInRange(int x, int y) {
```
Comment (isubo) [inline src/ua/griddynamics/tictactoe/Game.java]: Check only one value:
boolean isInRange(int x) {
....
}

**Review by isubo — APPROVED** [review]
